## Multiple Linear Regression

In [16]:
from warnings import filterwarnings
filterwarnings("ignore")

# Step 1 : Read the Data Set

In [17]:
import pandas as pd 
df=pd.read_csv("50_Startups.csv")
df

RND      ADMIN        MKT       STATE     PROFIT
0   165349.20  136897.80  471784.10    New York  192261.83
1   162597.70  151377.59  443898.53  California  191792.06
2   153441.51  101145.55  407934.54     Florida  191050.39
3   144372.41  118671.85  383199.62    New York  182901.99
4   142107.34   91391.77  366168.42     Florida  166187.94
5   131876.90   99814.71  362861.36    New York  156991.12
6   134615.46  147198.87  127716.82  California  156122.51
7   130298.13  145530.06  323876.68     Florida  155752.60
8   120542.52  148718.95  311613.29    New York  152211.77
9   123334.88  108679.17  304981.62  California  149759.96
10  101913.08  110594.11  229160.95     Florida  146121.95
11  100671.96   91790.61  249744.55  California  144259.40
12   93863.75  127320.38  249839.44     Florida  141585.52
13   91992.39  135495.07  252664.93  California  134307.35
14  119943.24  156547.42  256512.92     Florida  132602.65
15  114523.61  122616.84  261776.23    New York  129917.04
16   78013.11  121597.55  264346.06  California  126992.93
17   94657.16  145077.58  282574.31    New York  125370.37
18   91749.16  114175.79  294919.57     Florida  124266.90
19   86419.70  153514.11       0.00    New York  122776.86
20   76253.86  113867.30  298664.47  California  118474.03
21   78389.47  153773.43  299737.29    New York  111313.02
22   73994.56  122782.75  303319.26     Florida  110352.25
23   67532.53  105751.03  304768.73     Florida  108733.99
24   77044.01   99281.34  140574.81    New York  108552.04
25   64664.71  139553.16  137962.62  California  107404.34
26   75328.87  144135.98  134050.07     Florida  105733.54
27   72107.60  127864.55  353183.81    New York  105008.31
28   66051.52  182645.56  118148.20     Florida  103282.38
29   65605.48  153032.06  107138.38    New York  101004.64
30   61994.48  115641.28   91131.24     Florida   99937.59
31   61136.38  152701.92   88218.23    New York   97483.56
32   63408.86  129219.61   46085.25  California   97427.84
33   55493.95  103057.49  214634.81     Florida   96778.92
34   46426.07  157693.92  210797.67  California   96712.80
35   46014.02   85047.44  205517.64    New York   96479.51
36   28663.76  127056.21  201126.82     Florida   90708.19
37   44069.95   51283.14  197029.42  California   89949.14
38   20229.59   65947.93  185265.10    New York   81229.06
39   38558.51   82982.09  174999.30  California   81005.76
40   28754.33  118546.05  172795.67  California   78239.91
41   27892.92   84710.77  164470.71     Florida   77798.83
42   23640.93   96189.63  148001.11  California   71498.49
43   15505.73  127382.30   35534.17    New York   69758.98
44   22177.74  154806.14   28334.72  California   65200.33
45    1000.23  124153.04    1903.93    New York   64926.08
46    1315.46  115816.21  297114.46     Florida   49490.75
47       0.00  135426.92       0.00  California   42559.73
48     542.05   51743.15       0.00    New York   35673.41
49       0.00  116983.80   45173.06  California   14681.40

## step 2: Always Perform Basic Data Quality Checks

In [31]:
df.head()

RND      ADMIN        MKT       STATE     PROFIT
0  165349.20  136897.80  471784.10    New York  192261.83
1  162597.70  151377.59  443898.53  California  191792.06
2  153441.51  101145.55  407934.54     Florida  191050.39
3  144372.41  118671.85  383199.62    New York  182901.99
4  142107.34   91391.77  366168.42     Florida  166187.94

In [32]:
df.tail()

RND      ADMIN        MKT       STATE    PROFIT
45  1000.23  124153.04    1903.93    New York  64926.08
46  1315.46  115816.21  297114.46     Florida  49490.75
47     0.00  135426.92       0.00  California  42559.73
48   542.05   51743.15       0.00    New York  35673.41
49     0.00  116983.80   45173.06  California  14681.40

In [33]:
df.isna().sum()

RND       0
ADMIN     0
MKT       0
STATE     0
PROFIT    0
dtype: int64

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   RND     50 non-null     float64
 1   ADMIN   50 non-null     float64
 2   MKT     50 non-null     float64
 3   STATE   50 non-null     object 
 4   PROFIT  50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


In [34]:
df.nunique()

RND       49
ADMIN     50
MKT       48
STATE      3
PROFIT    50
dtype: int64

In [35]:
df.duplicated().sum()

0

In [24]:
df.describe()

RND          ADMIN            MKT         PROFIT
count      50.000000      50.000000      50.000000      50.000000
mean    73721.615600  121344.639600  211025.097800  112012.639200
std     45902.256482   28017.802755  122290.310726   40306.180338
min         0.000000   51283.140000       0.000000   14681.400000
25%     39936.370000  103730.875000  129300.132500   90138.902500
50%     73051.080000  122699.795000  212716.240000  107978.190000
75%    101602.800000  144842.180000  299469.085000  139765.977500
max    165349.200000  182645.560000  471784.100000  192261.830000

## Step 3 : Seperate X(RND, ADMIN, MKT) and Y(PROFIT)

### PROFIT ~ RND, ADMIN, MKT

PROFIT_pred = B0 + B1xRND + B2xADMIN + B3xMKT

X = [['RND','ADMIN','MKT']]
Y = [['PROFIT]]

In [39]:
X = df[['RND','ADMIN','MKT']]
Y = df[['PROFIT']]

In [40]:
X . head()

RND      ADMIN        MKT
0  165349.20  136897.80  471784.10
1  162597.70  151377.59  443898.53
2  153441.51  101145.55  407934.54
3  144372.41  118671.85  383199.62
4  142107.34   91391.77  366168.42

In [41]:
Y.head()

PROFIT
0  192261.83
1  191792.06
2  191050.39
3  182901.99
4  166187.94


### Step 4: Build a Linear Regression Model


In [45]:
from  sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X,Y)

LinearRegression()

###In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook.
On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.

In [46]:
model.intercept_

array([50122.19298987])

In [47]:
model.coef_

array([[ 0.80571505, -0.02681597,  0.02722806]])

In [49]:
X.columns

Index(['RND', 'ADMIN', 'MKT'], dtype='object')

In [50]:
Y.columns

Index(['PROFIT'], dtype='object')

### 
### PROFIT_pred = 50122.19 + 0.8057xRND - 0.0268xADMIN + 0.0272xMKT
## RND is having higest coefficient , Hence more i spend in RND more will be the profit


## Step 5: Evaluate the model


In [51]:
# R 2

model.score(X,Y)

0.9507459940683246

In [52]:
ypred = model.predict(X)
ypred[0:5]

array([[192521.25289008],
       [189156.76823227],
       [182147.2790962 ],
       [173696.70002553],
       [172139.51418327]])

In [53]:
Y.head()

PROFIT
0  192261.83
1  191792.06
2  191050.39
3  182901.99
4  166187.94

In [54]:
from sklearn.metrics import  mean_squared_error,mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error,r2_score


In [56]:
def evaluate_model(model, X, Y):
    ypred = model.predict(X)
    # Get all the metrics
    mse = mean_squared_error(Y, ypred)
    rmse = mse**(1/2)
    mae = mean_absolute_error(Y, ypred)
    mape = mean_absolute_percentage_error(Y, ypred)
    r2 = r2_score(Y, ypred)
    # Print all above results
    print(f'Mean Squared Error (MSE): {mse:.2f}')
    print(f'Root Mean Squared Error (RMSE) : {rmse:.2f}')
    print(f'Mean Absolute Error (MAE): {mae:.2f}')
    print(f'Mean Absolute Percentage Error (MAPE) : {mape*100:.2f} %')
    print(f'R2 score : {r2:.4f}')

In [57]:
evaluate_model(model, X, Y)

Mean Squared Error (MSE): 78417126.02
Root Mean Squared Error (RMSE) : 8855.34
Mean Absolute Error (MAE): 6471.45
Mean Absolute Percentage Error (MAPE) : 10.60 %
R2 score : 0.9507


## 
## Because above model has r2 score of 0.9507 > 0.8 hence above is a very good model


### 
### Step 6: Out of Sample Prediction


In [58]:
X.values[0:5]

array([[165349.2 , 136897.8 , 471784.1 ],
       [162597.7 , 151377.59, 443898.53],
       [153441.51, 101145.55, 407934.54],
       [144372.41, 118671.85, 383199.62],
       [142107.34,  91391.77, 366168.42]])

In [59]:
X.columns

Index(['RND', 'ADMIN', 'MKT'], dtype='object')

In [60]:
import seaborn as sns
import matplotlib.pyplot as plt

In [62]:
plt.figure(figsize=(10,5))
sns.scatterplot(data = df , x=("RND","ADMIN","MKT"),y="PROFIT")
plt.title("RND  Vs PROFIT ScatterPlot")
plt.show()

ValueError: Length of tuple vectors must match length of `data` when both are used, but `data` has length 50 and the vector passed to `x` has length 3.

<Figure size 1000x500 with 0 Axes>